### PR TITLE
(5/7) RFC79: Implement incremental upload for gene panel matrix

### PR DESCRIPTION
**Pull request by forus**
Please review https://github.com/cBioPortal/cbioportal-core/pull/40 next.

_Please note, to avoid the costly process of rebasing all pull requests (PRs), all feedback will be incorporated into the final PR._

### CODE REVIEW

File: scripts/importer/validateData.py [hunk 1]
```diff
@@ -3386,7 +3386,7 @@ def checkLine(self, data):
                 sample_ids_panel_dict[sample_id] = data[self.mutation_stable_id_index - 1]
                 # Sample ID has been removed from list, so subtract 1 position.
                 if data[self.mutation_stable_id_index - 1] != 'NA':
-                    if sample_id not in mutation_sample_ids:
+                    if mutation_sample_ids is not None and sample_id not in mutation_sample_ids:
```
[the remark this answers is not in the document]
Comment (forus) [reply scripts/importer/validateData.py:3389]: this check is switched off in the incremental upload mode as the rest of the cross-reference checks.

We can load the gene panel matrix without having the case list in this instance that specifies sample IDs.

File: src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java [hunk 2]
```diff
@@ -206,7 +206,7 @@ public void extractDataToImport(
             }
             throw new RuntimeException("Sample with stable id " + sampleIdStr + " is not found in the database.");
         }
-        createSampleProfile(sample);
+        ensureSampleProfileExists(sample);
```
[the remark this answers is not in the document]
Comment (forus) [reply src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:209]: the logic creates records in the `sample_profile` table even when `genePanelId` is `null`.

https://github.com/cBioPortal/cbioportal-core/blob/inc-gene-panel-matrix/src/main/java/org/mskcc/cbio/portal/dao/DaoSampleProfile.java#L62

I haven't changed anything in this respect.

File: src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java [hunk 3]
```diff
@@ -428,7 +457,7 @@ public void importData() throws IOException, DaoException {
                             mutations.put(mutation,mutation);
                         }
                         if(!sampleSet.contains(sample.getStableId())) {
-                            addSampleProfileRecord(sample);
+                            ensureSampleProfileExists(sample);
```
[the remark this answers is not in the document]
Comment (forus) [reply src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:460]: hm, a set with sample IDs that were already processed. I am not sure why the code not skipping these lines earlier.

[the remark this answers is not in the document]
Comment (forus) [reply src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java:460]: I've removed it after all here https://github.com/cBioPortal/cbioportal-core/pull/47/commits/13eb147cf5c8c607c72e4f21a0b81beecebc107b#diff-b3cccdbc78574935ab64e4df548ce86deb18cf7d2ca0967597862cc56a533c22L460

File: src/main/java/org/mskcc/cbio/portal/scripts/ImportGenePanelProfileMap.java [hunk 4]
```diff
@@ -85,6 +85,10 @@ public void run() {
                    "gene panel file" ).withRequiredArg().describedAs( "meta_file.txt" ).ofType( String.class );
             parser.accepts("noprogress", "this option can be given to avoid the messages regarding memory usage and % complete");
 
+            // supported by the uploader already. Added for uniformity, to do not cause error when upstream software uses this flag
+            parser.accepts("overwrite-existing",
+                    "Enables re-uploading molecular data that already exist for the given profile and sample.")
+                    .withOptionalArg().describedAs("overwrite-existing").ofType(String.class);
```
[the remark this answers is not in the document]
Comment (forus) [reply src/main/java/org/mskcc/cbio/portal/scripts/ImportGenePanelProfileMap.java:91]: see here https://github.com/cBioPortal/cbioportal-core/pull/47/commits/111f58e0f635c95d68bb5e2501c1a830f0749a87

**Review by pieterlukasse — APPROVED** [review]

**Review by haynescd — APPROVED** [review]
LGTM

**Pull request merged**